### PR TITLE
require form validation on submit for dynamic applications

### LINKED ISSUE #770: validate form when submitting dynamic application
currently form validity is not checked in the `:rems.workflow.dynamic/submit` command

### PR DESCRIPTION
**Pull request by opqdonut**
closes #770

### CODE REVIEW

File: src/cljs/rems/application.cljs [hunk 1]
```diff
@@ -205,10 +207,13 @@
                                  :description description}])
      (save-application app description app-id catalogue-ids items licenses
                        (fn [resp]
+                         (prn :RESP resp)
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: Why prn?

File: src/cljs/rems/application.cljs [hunk 2]
```diff
@@ -403,11 +408,17 @@
...
+  [application msgs]
+  (let [titles-by-id (into {}
+                           (concat
+                            (for [f (:items application)]
+                              [[:item (:id f)] (:title (localize-item f))])
+                            (for [f (:licenses application)]
+                              [[:license (:id f)] (:title (localize-item f))])))]
+    (into [:ul]
+          (for [m msgs]
+            (do (prn :MSG m)
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: Why prn?

File: src/cljs/rems/application.cljs [hunk 3]
```diff
@@ -403,11 +408,17 @@
...
 (defn- format-validation-messages
-  [msgs language]
-  (into [:ul]
-        (for [m msgs]
-          ;; TODO this is broken, :title is no longer provided
-          [:li (text-format (:key m) (get-in m [:title language]))])))
+  [application msgs]
+  (let [titles-by-id (into {}
+                           (concat
+                            (for [f (:items application)]
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: `item` instead of `f`?

File: src/cljs/rems/application.cljs [hunk 4]
```diff
@@ -403,11 +408,17 @@
...
+                              [[:item (:id f)] (:title (localize-item f))])
+                            (for [f (:licenses application)]
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: `license` instead of `f`?

File: src/cljs/rems/application.cljs [hunk 5]
```diff
@@ -141,7 +141,7 @@
                            {:application-id application-id}
                            {:catalogue-items catalogue-items}))}))
 
-(defn- submit-application [app description application-id catalogue-items items licenses]
+(defn- submit-application [app description application-id catalogue-items items licenses prev-validation]
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: Why the name `prev-validation`?

File: src/cljs/rems/application.cljs [hunk 6]
```diff
@@ -154,7 +154,9 @@
                              (rf/dispatch [::enter-application-page application-id]))
                          (rf/dispatch [::set-status {:status :failed
                                                      :description description
-                                                     :validation (:validation resp)}])))
+                                                     :validation prev-validation
+                                                     ;; this error is only shown if prev-validation was nil
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: Hmm?

File: src/cljs/rems/application.cljs [hunk 7]
```diff
@@ -154,7 +154,9 @@
...
+                                                     :error {:status-text (pr-str (:errors resp))}}])))
```
Comment (Macroz) [inline src/cljs/rems/application.cljs]: Why `pr-str` the errors and not localized?

File: src/clj/rems/db/applications.clj [hunk 8]
```diff
@@ -944,7 +940,9 @@
   (not (nil? (users/get-user-attributes userid))))
 
 (defn- valid-form-inputs? [application-id]
-  (= :valid (form-validation/validate (get-form-for application-id))))
+  ;; TODO a bit ugly:
+  (let [user-id (:applicantuserid (first (db/get-applications {:id application-id})))]
```
Comment (Macroz) [inline src/clj/rems/db/applications.clj]: Maybe there could be a better (named) helper function for this.

File: src/cljs/rems/application.cljs (hunk 2 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: oops

File: src/cljs/rems/application.cljs (hunk 5 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: it's the validation from the previous save. or would you prefer just `validation`?

File: src/cljs/rems/application.cljs (hunk 6 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: if there are validation errors, the status-modal shows them and not the `:error`

File: src/cljs/rems/application.cljs (hunk 7 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: Yeah eventually we should have localizations for all the errors that the dynamic command endpoint can return. However for now I think this is better than the previous "Failed: ()" message.

File: src/clj/rems/db/applications.clj (hunk 8 above)
Comment (opqdonut) [reply src/clj/rems/db/applications.clj]: suggestions? `(applicant-of-application application-id)`?

File: src/cljs/rems/application.cljs (hunk 5 above)
Comment (Macroz) [reply src/cljs/rems/application.cljs]: Just wondering why to submit an application you need to know the previous validation.

File: src/clj/rems/db/applications.clj (hunk 8 above)
Comment (Macroz) [reply src/clj/rems/db/applications.clj]: Or `get-applicant-of-application`?

File: src/cljs/rems/application.cljs (hunk 7 above)
Comment (Macroz) [reply src/cljs/rems/application.cljs]: Maybe you could add these missing features as separate bullets into the issue on the board so we can continue on this task in later PRs?

File: src/cljs/rems/application.cljs (hunk 6 above)
Comment (Macroz) [reply src/cljs/rems/application.cljs]: I think this part can be improved when we split the the different old actions to new files like the new actions are now.

File: src/cljs/rems/application.cljs (hunk 5 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: Save returns the validation errors, but dynamic submit just returns `{:errors [:form-not-valid]}`

I don't think we should add validation messages to the return schema of dynamic /command

File: src/cljs/rems/application.cljs (hunk 7 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: Fair enough

File: src/cljs/rems/application.cljs (hunk 5 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: I'm not sure I get what you mean.

File: src/cljs/rems/application.cljs (hunk 7 above)
Comment (opqdonut) [reply src/cljs/rems/application.cljs]: Created #798

**Pull request merged**